### PR TITLE
feat: support escalation end events

### LINKED ISSUE #10684: Catch an escalation on sub process boundary event
**Is your feature request related to a problem? Please describe.**
Escalations thrown from `subProcess` can be handled in the workflow by catching it on boundary events.

image

**Describe the solution you'd like**
* the escalation event contains an `escalationCode` (any string), **optional**
* the escalation is caught on the boundary event if the `escalationCode` in the BPMN is equal to the thrown `escalationCode`
* the escalation is caught on the boundary event if the `escalationRef` is omitted or the `escalationCode` of referenced escalation is omitted
* the boundary can be attached to the corresponding subProcess
* the boundary can be interrupting or non-interrupting
* The escalation should be caught by the "nearest enclosing parent activity"

**Out of scope**:
- Boundary event on CallActivity

### LINKED ISSUE #10686: Throw escalation end event
**Is your feature request related to a problem? Please describe.**
I want to use escalation end events in my process to communicate between parent and sub-processes.

**Describe the solution you'd like**
I can throw an escalation using an end event.

image

**Additional context**

- We can handle this in the `EndEventProcessor`

Sample XML:
```xml
<bpmn:endEvent id="Event_0vica91">
  <bpmn:incoming>Flow_0zt0ojv</bpmn:incoming>
  <bpmn:escalationEventDefinition id="EscalationEventDefinition_0w6r586" escalationRef="Escalation_3rhpl03" />
</bpmn:endEvent>

<bpmn:escalation id="Escalation_3rhpl03" name="Escalation_3rhpl03" escalationCode="escalationCode" />
```

### LINKED ISSUE #10690: Catch an escalation in an event subprocess start event
**Is your feature request related to a problem? Please describe.**
Thrown escalations can be handled in the workflow by catching it on an event subprocess.

image

**Describe the solution you'd like**
* the escalation event contains an `escalationCode` (any string), **optional**
* If `escalationRef` is omitted or `escalationCode` of referenced escalation is omitted, the event start event is triggered by any escalation event.
* In case an `escalationRef` is set, the start event is only triggered by escalation events with the defined `escalationCode`.
* The escalation code of the start event must be unique across the event sub-processes of the same scope. (part of #10688 validations)
* If a start event has no escalationRef or escalationCode of referenced escalation then another event sub-process with an escalation start event in **the same scope** is not supported. (part of #10688 validations)
* The escalation must be caught by the "nearest enclosing parent activity"

### LINKED ISSUE #10763: Catch escalation on a call activity boundary event
## Description
Escalations thrown from a called activity can be handled in the workflow by catching it on a boundary event.
image

**Describe the solution you'd like**
- The escalation event contains an `escalationCode` (any string), **optional**
- The escalation is caught on the boundary event if the `escalationCode` in the BPMN is equal to the thrown `escalationCode`
- The escalation is caught on the boundary event if the `escalationRef` is omitted or the `escalationCode` of referenced escalation is omitted
- The boundary event can be attached to the call activity
- The boundary event can be interrupting or non-interrupting
* The escalation should be caught by the "nearest enclosing parent activity"

### PR DESCRIPTION
**Pull request by lzgabel**
## Description
I can throw an escalation using an end event.

* Escalations thrown from a called activity or sub process can be handled by catching on a boundary event.
* Throw escalations can be handled by catching on an event subprocess.

## Related issues
closes #10686
closes #10684
closes #10690
closes #10763

## Definition of Done

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by lzgabel** [comment]
Hi @remcowesterhoud. Please take a moment to review this feature when you have time. 🙇

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:162]: 🙃 Please note that this isn't a must. This code is fine and readable. Personally I would've used recursion for this:

```java
      return findEscalationCatchEventInProcess(escalationCode, process, instance)
          .or(
              () -> {
                final ElementInstance parentElementInstanceKey =
                    elementInstanceState.getInstance(instanceRecord.getParentElementInstanceKey());
                return findEscalationCatchEvent(escalationCode, parentElementInstanceKey);
              });
```

This would also involve changing from a `while`-loop to an `if`-statement

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:181]: 🙃 Again, not necessary, but this could be achieved with a recursive call as well 😄

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:198]: ❌ We should change this method to not care about whether or not an event is a start event or a boundary event. We can quite easily achieve this with a simple stream. Using this stream would make this check obsolete so we can remove it.

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:229]: ❌ We can replace this all with a simple stream. We don't have to care if an event is a start event or a boundary event. As long as we are searching for the events in the correct order, meaning from the scope closest to the throw event to the scope farthest away from the throw event, we can return the first event we find regardless of its event type.

```java
    final Optional<ExecutableCatchEvent> catchEvent =
        element.getEvents().stream()
            .filter(ExecutableCatchEvent::isEscalation)
            .filter(
                event -> {
                  final var eventEscalationCode = event.getEscalation().getEscalationCode();
                  return eventEscalationCode.capacity() == 0
                      || eventEscalationCode.equals(escalationCode);
                })
            .findFirst();
```

This should be enough to find the correct catch event. For readability's sake we could pull the 2nd filter to a separate method to make it more descriptive. Something along the lines of `matchesEscalationCode(event, escalationCode)`.

If the optional contains a catchEvent we can set the tuple and return an optional of it. Otherwise we must return an empty optional.

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:244]: ❌ This method becomes obsolete if you apply the previous suggestions.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks again @lzgabel! For the most part the changes look great! 🚀 

I have added some suggestions, mainly to the way you are currently finding the catch event. Please have a look and let me know what you think!

I'd also like to make one more suggestion here as I can't comment on unchanged code:

❌ We should rename the methods in the `CatchEventAnalyzer` that are related to finding an error event accordingly. Currently they read like they are finding any catch event. Now that this class find multiple types of events we should change this.
E.g. `findCatchEvent` -> `findErrorCatchEvent`

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java
Comment (lzgabel) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:263]: > We should also add a testcase where we catch the escalation of the child process in an event sub-process instead of with a boundary event.

Hi @remcowesterhoud. It seems that this [test case](https://github.com/camunda/zeebe/pull/10914/files?diff=unified&w=0#diff-a2b6a08f3686d9b29d02a506fd9e57b5de1ed5f42bc09bc5efbf1824a4eb67b3R111-R136) already covers this scenario. 🤔  WDYT?

Comment (remcowesterhoud) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:263]: Hey @lzgabel. I think it'd be good to have the combination of a call activity + an event sub-process. This is a slightly different scenario. I believe it'll work with the current code but it's good to test it.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/escalation/EscalationEventTest.java:421]: 🔧 I think there was a misunderstanding between us about the testcase I was missing. What I meant was this:

We have a parent process with a call activity. The call activity does **not** have a boundary event. Instead the parent process contains an event sub process which will catch the escalation.
The child process that is called by the call activity will simply throw an escalation.

E.g.
Parent:
image

Child:
image


This is different from what you did here, as the event sub process is part of the child process in the current  case.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnEventPublicationBehavior.java:124]: 🙃 This is somewhat nitpicky, but I would prefer renaming this from `activated` to `context`. This behavior shouldn't have to care at what stage this method is called. If we want to throw an escalation event during completion of an element we can use the same method, though the name doesn't make sense in that case.

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:153]: 🔧 Let's add a comment to this block similar to what you had before:

`find in parent process instance if exists`

File: engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java:172]: 🔧 Let's add a comment to this bock similar to what you had before:

`find in parent scope if exists`

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks for all the changes @lzgabel! I only have a few more comments left, after this I can go ahead and merge it! 🥇

**Comment by remcowesterhoud** [comment]
The flaky test that currently fails the build has been fixed with #10938 so we can ignore this here.

**Comment by lzgabel** [comment]
Hi @remcowesterhoud. I've fixed all review comments. Please check this out, I will squash the commit after this round of review. 🙇

**Review by remcowesterhoud — APPROVED** [review]
Great work @lzgabel ❤️ 

Go ahead and squash it. I'll merge it when you're done.

**Comment by lzgabel** [comment]
Done. 

Hi @remcowesterhoud. Please check this out. 🙇

**Comment by remcowesterhoud** [comment]
Awesome! Thank you 🙇 

bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Pull request merged**
